### ISSUE TITLE
Allow multiple output files for each command

### ISSUE THREAD

**Issue by wlandau** [ropensci/drake#283]
Contemplated in #232 and #257. I just saw that [`snakemake`](https://bitbucket.org/snakemake/snakemake/) allows any number of output files, and I think `drake` should try to catch up.

```r
drake_plan(
  data = get_data(),
  tmp = {
    saveRDS(data, file_out("data.rds"))
    saveRDS(summary(data), file_out("summary.rds"))
  }
  my_plot = make_plot(file_in("data.rds", "summary.rds"))
)
```

From #232, it is not enough to just hack at the problem. I think it will take a rigorous design specification to plan clearly beforehand.

We may need to include a transformed version of the network graph. The current graph links together all the imported and generated objects/files. For the desired behavior, I think we need to collapse that into a dependency graph of job IDs. I was resisting this in #232.

**Comment by krlmlr** [comment]
For R objects, the analogy is unpacking a named list: A command that returns multiple objects, of which separate components can be used by downstream targets.

**Comment by wlandau** [comment]
I agree that we will need to unpack all the outputs if we have multiple. Could #285 will make it easier? `Drake` could handle a dependency graph of objects and files, while the external scheduler could have its own dependency graph of jobs. The reason I bring up different graphs again is that the scheduler's graph may be different anyway when we start thinking about multidimensional knapsack and transient workers with multiple jobs each.

**Comment by krlmlr** [comment]
Maybe we want to switch to bipartite graphs internally, with two kinds of nodes: commands and targets? That would support M:N relationships between targets. The scheduler would operate on a simplified version where the target nodes are collapsed.

The other situation where the schedule graph is different from the dependency graph is in the context of a DSL that supports delayed expansion: A node/edge in the dependency graph can correspond to many nodes/edges in the schedule graph.

Does the terminology make sense: "schedule graph" vs. "dependency graph"?

**Comment by krlmlr** [comment]
I also feel that it would pay off to consult a text book about scheduling first. Just to make sure we don't reinvent too many wheels. I'm not familiar with the literature, though.

**Comment by wlandau** [comment]
Re: graphs: yes, I understand and support both the concept and the choice of terminology. Incidentally, `drake` is already (clumsily) using two graphs anyway: `config$graph` and `config$execution_graph`. The former is a dependency graph that does not get modified, and the latter is a schedule graph, copied from `config$graph` and trimmed over the course of staged parallelism.

And yes, we should do our research when it comes to scheduling. And I think we should maintain a design specification. I have a stub in the [`scheduler` branch of `drake`](https://github.com/ropensci/drake/tree/scheduler).

**Comment by wlandau** [comment]
I thought more about how to solve this one:

- [ ] Allow multiple files in the `file_out()` function itself.
- [ ] Stop forcing `file_out()` files to be target names. There will be one return value per command, which will no longer be a file.
- [ ] Augment `config$graph` to include all the `file_out()` files alongside the return value of each command. Here, the `file_out()` files and the return value share all their dependencies and are conditionally independent. They also share a job ID as a `igraph` attribute, which is just the name of the return value (from the workflow plan data frame).
- [ ] Create `config$schedule`, which is just `config$graph` with the `file_out()` files removed (file nodes with non-null job ID `igraph` attributes).
- [ ] Make all the parallel backends use `config$schedule` instead of `config$graph` to spin up jobs.
- [ ] Create a function to get all the `file_out()` files of a target or job.
- [ ] Somewhere in `conclude_build()`, after the return value of the command is stored, check for `file_out()` files and store them similarly.
- [ ] Modify the "file" trigger to check for all the `file_out()` files of a target.

I think I covered it. I will be chipping away at this issue in the [i283 branch](https://github.com/ropensci/drake/tree/i283), especially after the CRAN release of version 5.1.0 this Thursday.

**Comment by krlmlr** [comment]
Does this plan also provision for commands with multiple output objects?

We could also use a bipartite graph to model this:

1. Target nodes
2. Command nodes

Currently, commands are modeled as edges. The new model adds more nodes but reduces the number of edges, and may be a simpler change.

**Comment by wlandau** [comment]
Currently, commands have only one non-file return value. When the DSL changes this, we can update the internals accordingly. It should be relatively straightforward.

Having worked through some of this in the `i283` branch, I think something similar to a bipartite graph would increase performance relative to a partitioning based on `igraph` attributes. Adding a bipartite graph would leave us with 3 graphs:

1. a dependency graph of just targets and imports
2. a schedule graph of the jobs
3. a bipartite graph to map between targets/imports and jobs

I think this is reasonable as long as we are okay with having so many graphs.

**Comment by krlmlr** [comment]
Aren't these three graphs just different views/transformations of the same graph?

**Comment by wlandau** [comment]
Yes they are, and the duplicated information does bother me. But in `drake`'s internals, I do not see any situation where we would actually need the union of the three (or any pair, for that matter). Any time we need to learn something from the graph, we will ultimately take one of these views, and selecting the right view repeatedly from the larger graph could decrease performance.

**Comment by wlandau** [comment]
But then again, maybe we should worry about elegance first. Maybe selecting `igraph` nodes by attribute is fast.

**Comment by krlmlr** [comment]
Agree that we should maintain one single "master" data structure from which all helper data structures can be derived (on the fly or by some preprocessing).?

**Comment by wlandau** [comment]
Hmm... I am trying again to implement this, and we actually need the dependency view on its own to construct the schedule view. In that sense, I am not actually sure we should glue the two views together if they are created separately and we are just going to separate them all over again.

**Comment by wlandau** [comment]
From f14645082afa38d0b11e5be2201cac585c7b38e3, I found that with the right attributes, the schedule falls right out of the dependency graph super easily. If there turns out to be added implementation-level value of an overarching master graph, we can certainly combine the two graphs together.

**Comment by wlandau** [comment]
My work on the `i283` branch so far seems to be making `drake`'s internals a little more complicated, a little slower, and a bit more brittle, so I want to think out loud here about a different approach to the issue. Is there any real reason we have to include concomitant `file_out()` files explicitly in a graph? What if `config$graph` is the schedule and we just modify file triggers and target storage to account for dependencies? We would need to do a bit of extra work to construct the dependency graph from the schedule, but this should only become relevant in `dataframes_graph()` (for `vis_drake_graph()`). I like the design of the triggers much more than most aspects of `drake`'s internals. If we focus our efforts there instead of the graph, we may have better luck both here and with #233.

**Comment by wlandau** [comment]
Or maybe the DSL could handle this with `pack()`/`unpack()` (#304).

**Comment by wlandau** [comment]
Update: the [i283 branch](https://github.com/ropensci/drake/tree/i283) appears to make this work for me for a small test case. There is a lot of testing/debugging work left to do, and I would definitely appreciate a reviewer for the pull request when the time comes.

I am still wondering if this is the right solution. I worry that the code base got more complicated than it needs to be.

**Comment by krlmlr** [comment]
Happy to review if time permits. Would you like to open a PR for discussion?

**Comment by wlandau** [comment]
Thanks, Kirill. I put my work on this aside for a while because it complicated the internals and wreaked havoc on the tests, but I suppose a PR just for the sake of discussion would help.

**Comment by wlandau** [comment]
#350 did not work out. I will attempt again after #440 is totally solved. Then, the solution will be cleaner, faster, and more elegant.

Finding the right data structure is more than half the battle.

**Comment by wlandau** [comment]
Both #350 and #457 rely on deep fundamental infrastructure changes that make the internals more complicated. I am starting to wonder if there is a less invasive route.

**Comment by wlandau** [comment]
I am attempting to solve this again with a minimally invasive approach. The solution will proceed step by step with small with one small PR at a time, and there will be no major additions or changes to internal data structures.

Let's keep main parts of `config$graph` just the way they are: no fooling around with nodes or edges, and no holding on to two graph representations. To keep track of output files, let's use graph attributes instead. Then, each target's list of output files makes its way to `drake_meta()`.

One downside is that we won't see the `file_out()` files in `vis_drake_graph()` and friends. However, I think this approach is still worth it: it is less error prone, and `drake` will run faster because it does not need to create a whole new complicated graph representation. 

Our to-do list is much shorter than in the last attempt (https://github.com/ropensci/drake/issues/283#issuecomment-374445808):

- [ ] Carry over `file_out()` files to `drake_meta()` lists.
- [ ] Stop forcing `file_out()` files to be target names in the workflow plan. There will be one return value per command, which will no longer be a file (once the old file API is removed).
- [ ] Allow multiple files in the `file_out()` function itself.
- [ ] Somewhere in `conclude_build()`, after the return value of the command is stored, store the fingerprints and metadata of all the `file_out()` files too.
- [ ] Modify the "file" trigger and `meta$file` to check all the `file_out()` files of a target.

**Comment by wlandau** [comment]
> Then, each file_out() will be removed in a way that connects each upstream node directly to each downstream node.

Here is what I mean.

```r
library(igraph)
opt <- igraph::igraph_opt("return.vs.es")
on.exit(igraph::igraph_options(return.vs.es = opt))
igraph::igraph_options(return.vs.es = FALSE)
graph <- make_ring(10, directed = TRUE)
V(graph)$name <- as.character(1:10)
vertex <- "7"
from <- unique(unlist(adjacent_vertices(graph, vertex, mode = "in")))
to <- unique(unlist(adjacent_vertices(graph, vertex, mode = "out")))

plot(graph)
graph <- graph + edges(expand.grid(from = from, to = to))
plot(graph)
graph <- delete_vertices(graph, v = vertex)
plot(graph)
```

**Comment by wlandau** [comment]
This solution is getting really messy too. It's not as simple as keeping better track of the `file_out()` files in the metadata. I will revert the earlier PR and keep the branch. This feature probably won't be available until `drake` 6.0.0 at least.

**Comment by wlandau** [comment]
On the other hand, other parts of the code clean up super nicely: for example, `clean.R`. And with a recent fix, the tests are starting to work. It may be strange to store file inputs and file outputs as `igraph` attributes in a schedule of jobs, but it's the least invasive thing I can do. I will proceed.

**Comment by wlandau** [comment]
Fixed via #469!
